library/Core/src/Juvix/Core/IR/Types/Base.hs

```diff
@@ -122,38 +125,101 @@ type GlobalAllWith (c :: * -> Constraint) ty ext primTy primVal =
     c primVal,
     TermAll c ext primTy primVal,
     ElimAll c ext primTy primVal,
+    ValueAll c ext primTy primVal,
+    NeutralAll c ext primTy primVal,
     PatternAll c ext primTy primVal
   )
 
-data DatatypeWith ty primTy primVal
+data DatatypeWith ty ext primTy primVal
```

Comment:
let's consider splitting this into `DatatypeWithPreTypecheck` and `DatatypeWithPostTypecheck` so we don't have the duplicate `dataArgs` / `nfDataArgs` fields?

Reply:
(or figure out how to use the `ty` argument)

Reply:
Andy and I discussed further, we will have a data type with `Value'` and a separate with with `Term'`.